### ISSUE TITLE
Project 2 Level 6 (test6_7)

### ISSUE THREAD

**Issue by crys-22** [nus-cs2030/2021-s1#411]
# Description 
Did anyone get stuck at this testcase & manage to clear it? I'm not sure how to continue from this because I can't really debug also because idk what are the inputs and somehow my code works fine for the public testcases... I can tell that 15 is placed to wait for self-check 4 instead of 3, but i'm not sure why this is happening because I do check for which self-check has the closest next avail time & it should be 3...?? :( pls send help! 

## Screenshots (if any): 
![image](https://user-images.githubusercontent.com/71073099/99415800-25d94900-2933-11eb-82bd-5400e22654de.png)

**Comment by waynetee** [comment]
observe that in your code 15 is served after 16 and 17 even though he arrived earlier. You should use a shared queue for the self checkout counters to ensure that earlier customers get served first, hope this helps

**Comment by crys-22** [comment]
hi, thanks for replying! I do use a shared queue! In ArriveEvent, I will assign the customer to the self-check customer with the nearest next avail time. In this case, it should be 3, but I'm not sure why it makes 4 become the one with the nearest next avail time..... i considered that it might be an updating issue as well, but if it was, perhaps it would have been reflected on the public testcases as well?? i'm not sure...

**Comment by fans2619** [comment]
Hi, I'm not sure how you can guarantee what self-checkout counter will serve an arriving customer when the customer arrives. However, from what you said, it doesn't sound like a shared queue indeed. You need a shared queue according to the fact that all the customers are waiting for all the self-checkout counters. However, I don't think a customer will know which self-checkout counter will serve him when he is arriving because the all the service durations are random.

**Comment by qiaoningg** [comment]
Not sure if you're using static to achieved the "shared queue" but if you're not, you can try that.

**Comment by b-leong** [comment]
You might want to check that you assigned the correct self checkout server to the customers (earliest next available time) and your updating also, maybe more print statements to check, because it seems like 15 is wrongly assigned to self check 4

**Comment by crys-22** [comment]
> You might want to check that you assigned the correct self checkout server to the customers (earliest next available time) and your updating also, maybe more print statements to check, because it seems like 15 is wrongly assigned to self check 4

![image](https://user-images.githubusercontent.com/71073099/99478348-30c1c700-298f-11eb-84d1-3ff777567c94.png)
I printed out the nextAvailTime of 3 vs the best option (earliest one which happens to be 4), and i realised 4 is really the earliest next available time, so now im confused as to why 15 gets served by 3 shouldnt it be 4 because 4 is the earliest timing

**Comment by b-leong** [comment]
Whats the done event time for cust 14

**Comment by b-leong** [comment]
Most likely when you checked the next available time when customer 15 wait event, you compared the next available time of the servers without accounting for the people ahead in queue. I suspect that you assigned 15 to self 4 because the next avail time was checked before it added the service time to serve customer 14 as the serve event for cust 14 is after the wait event for cust 15

**Comment by crys-22** [comment]
thanks everyone for your insights!!! I managed to resolve this le!!! in case anyone wants to know how:
yes, i had a shared queue but as said above i shouldnt have assigned a specific sc to it during ArriveEvent, and instead do that in the waitevent such that that server is passed to the next ServeEvent. 
My main issue was that for the ServeEvent, under the condition that there is a clash, i did not account for self-checks. I resolved this by including the condition for self-checks, and passed in a new server to the new ServeEvent returned.
